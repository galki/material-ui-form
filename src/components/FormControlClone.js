// @flow

import React from 'react'
import _ from 'lodash'

import { FormControl, FormHelperText, FormLabel } from 'material-ui/Form'
import { InputLabel } from 'material-ui/Input'

import propNames from '../propNames'


function getErrorAndHelperText(field: Object): Object {
  let helperText: ?string
  let isError: boolean = false
  if (!_.isEmpty(field) && field.validations.length > 0) {
    helperText = field.validations[0].message
    isError = true
  }
  return { helperText, isError }
}

type Props = {
  field?: Object,
  formControlComp: Object,
  onValueChange: Function,
  onConstruct: Function,
};

type State = {
  helperText: ?string,
  isError: boolean,
  value: mixed,
};

export default class FormControlClone extends React.Component<Props, State> {
  static defaultProps = {
    field: {},
  }

  // eslint-disable-next-line react/sort-comp
  helperText: string
  name: string

  constructor(props: Object) {
    super(props)

    const { error, required } = props.formControlComp.props

    let name
    let value
    let helperText
    let isError = error

    React.Children.forEach(props.formControlComp.props.children, (child) => {
      if (child.type === FormHelperText) {
        helperText = String(child.props.children)
        this.helperText = helperText
      } else if (child.type !== FormLabel
        && child.type !== InputLabel
        && child.props.name !== undefined
        && child.props.value !== undefined
      ) {
        name = child.props.name // eslint-disable-line prefer-destructuring
        value = child.props.value // eslint-disable-line prefer-destructuring
      }
    })

    if (props.formControlComp.type !== FormControl
      || name === undefined
      || value === undefined
    ) {
      throw new Error('invalid FormControl control children')
    }

    if (props.field.value === undefined) {
      const validatorsPropName = propNames.FIELD_VALIDATORS
      props.onConstruct({
        name,
        value,
        required,
        [validatorsPropName]: props.formControlComp.props[validatorsPropName],
      })
    } else {
      value = props.field.value // eslint-disable-line prefer-destructuring
      if (!_.isEmpty(props.field) && props.field.validations.length > 0) {
        const fieldError = getErrorAndHelperText(props.field)
        helperText = fieldError.helperText // eslint-disable-line prefer-destructuring
        isError = fieldError.isError // eslint-disable-line prefer-destructuring
      }
    }

    this.name = name
    this.state = {
      helperText,
      isError,
      value,
    }
  }

  // eslint-disable-next-line
  UNSAFE_componentWillReceiveProps(nextProps: Object) {
    if (!_.isEmpty(nextProps.field)) {
      const { helperText, isError } = getErrorAndHelperText(nextProps.field)
      this.setState({
        helperText,
        isError,
        value: nextProps.field.value,
      })
    }
  }

  onChange = (event: SyntheticInputEvent<Element>) => {
    const { value } = event.target
    this.setState({ isError: false, helperText: this.helperText, value })
    this.props.onValueChange(this.name, value, true)
  }

  render() {
    const { formControlComp, formControlComp: { props } } = this.props

    let hasHelperText = false
    const children = React.Children.map(props.children, (child) => {
      // label
      if (child.type === FormLabel || child.type === InputLabel) {
        return child
      }
      // helper text
      if (child.type === FormHelperText) {
        hasHelperText = true
        return React.cloneElement(child, {
          children: this.state.helperText,
        })
      }
      // field
      return React.cloneElement(child, {
        onChange: child.props.onChange || this.onChange,
        value: this.state.value,
      })
    })
    // support for dynamic helper text
    if (!hasHelperText && this.state.helperText !== undefined) {
      children.push(<FormHelperText key={1}>{this.state.helperText}</FormHelperText>)
    }

    return (
      React.cloneElement(formControlComp, {
        error: this.state.isError,
        children,
      })
    )
  }
}
