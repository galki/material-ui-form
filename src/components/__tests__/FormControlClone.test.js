import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import {
  FormLabel,
  FormControl,
  FormControlLabel,
  FormHelperText,
} from 'material-ui/Form'
import Radio, { RadioGroup } from 'material-ui/Radio'
import Input, { InputLabel } from 'material-ui/Input'
import Select from 'material-ui/Select'
import { MenuItem } from 'material-ui/Menu'

import FormControlClone from '../FormControlClone'


const field = {
  isPristine: true,
  isRequired: null,
  pristineValue: null,
  validations: [],
  validators: [],
  value: undefined,
}

describe('<FormControlClone>:<Select>', () => {
  const formControlElement = (
    <FormControl required>
      <InputLabel htmlFor="age-helper">Age</InputLabel>
      <Select value="" name="age">
        <MenuItem value=""><em>Please select your age ...</em></MenuItem>
        <MenuItem value={10}>Teens</MenuItem>
        <MenuItem value={20}>Twenties</MenuItem>
        <MenuItem value={30}>Thirties</MenuItem>
        <MenuItem value="40+">Fourties +</MenuItem>
      </Select>
      <FormHelperText>Some important helper text</FormHelperText>
    </FormControl>
  )

  const wrapper = shallow(
    <FormControlClone
      field={field}
      formControlElement={formControlElement}
      onConstruct={jest.fn()}
      onValueChange={jest.fn()}
    />
  )

  it('should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  // TODO: how to simulate on children?
  // it('should handle onChange events of children', () => {
  //   const value = '10'
  //   const event = { target: { value } }
  //   wrapper.find(FormControlClone).simulate('change', event)
  //   expect(wrapper.state()).toMatchObject({
  //     isError: false,
  //     value,
  //   })
  //   expect(wrapper.instance().props.onValueChange).toHaveBeenCalled()
  // })
})

describe('<FormControlClone>:<RadioGroup>', () => {
  const formControlElement = (
    <FormControl
      component="fieldset"
      required
    >
      <FormLabel component="legend">
        RadioGroup FormControl
      </FormLabel>
      <RadioGroup
        name="certainty"
        value=""
      >
        <FormControlLabel
          value="high"
          control={<Radio />}
          label="I swear"
        />
        <FormControlLabel
          value="soso"
          control={<Radio />}
          label="Probably"
        />
        <FormControlLabel
          value="low"
          control={<Radio />}
          label="Maybe"
        />
      </RadioGroup>
    </FormControl>
  )

  const wrapper = shallow(
    <FormControlClone
      field={field}
      formControlElement={formControlElement}
      onConstruct={jest.fn()}
      onValueChange={jest.fn()}
    />
  )

  it('should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
