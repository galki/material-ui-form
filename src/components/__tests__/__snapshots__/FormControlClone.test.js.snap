// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormControlClone>:<RadioGroup> should render 1`] = `
<WithStyles(FormControl)
  component="fieldset"
  required={true}
>
  <WithStyles(FormLabel)
    component="legend"
    key=".0"
  >
    RadioGroup FormControl
  </WithStyles(FormLabel)>
  <RadioGroup
    key=".1"
    name="certainty"
    onChange={[Function]}
    value=""
  >
    <WithStyles(FormControlLabel)
      control={<WithStyles(Radio) />}
      label="I swear"
      value="high"
    />
    <WithStyles(FormControlLabel)
      control={<WithStyles(Radio) />}
      label="Probably"
      value="soso"
    />
    <WithStyles(FormControlLabel)
      control={<WithStyles(Radio) />}
      label="Maybe"
      value="low"
    />
  </RadioGroup>
</WithStyles(FormControl)>
`;

exports[`<FormControlClone>:<Select> should render 1`] = `
<WithStyles(FormControl)
  required={true}
>
  <WithStyles(InputLabel)
    htmlFor="age-helper"
    key=".0"
  >
    Age
  </WithStyles(InputLabel)>
  <WithStyles(Select)
    key=".1"
    name="age"
    onChange={[Function]}
    value=""
  >
    <WithStyles(MenuItem)
      value=""
    >
      <em>
        Please select your age ...
      </em>
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      value={10}
    >
      Teens
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      value={20}
    >
      Twenties
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      value={30}
    >
      Thirties
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      value="40+"
    >
      Fourties +
    </WithStyles(MenuItem)>
  </WithStyles(Select)>
  <WithStyles(FormHelperText)
    key=".2"
  >
    Some important helper text
  </WithStyles(FormHelperText)>
</WithStyles(FormControl)>
`;
